### ISSUE TITLE
tf.print  can't print log in jupyter notebook. 

### ISSUE THREAD

**Issue by lyhue1991** [tensorflow/tensorflow#24686]
When I tried to use tf.print() or get the log_device_placement information in google colabratory. I just can't get the print output. I tried in my own jupyter, and I find the print log is in the cmd.exe windows. This behavior is very strange and I think it should be fixed.  The code is like below.

`python
import tensorflow as tf
print(tf.GIT_VERSION, tf.VERSION)
a = tf.constant([1.0, 2.0, 3.0, 4.0, 5.0, 6.0], shape=[2, 3], name='a')
b = tf.constant([1.0, 2.0, 3.0, 4.0, 5.0, 6.0], shape=[3, 2], name='b')
c = tf.matmul(a, b)
# Creates a session with log_device_placement set to True.
sess = tf.Session(config=tf.ConfigProto(log_device_placement=True))
# Runs the op.
print(sess.run(c))
`

**Comment by ymodak** [comment]
By adding tf.logging.set_verbosity(tf.logging.INFO in your script you should be able to see the logging in the ipython console.

**Comment by lyhue1991** [comment]
> By adding **tf.logging.set_verbosity(tf.logging.INFO)** in your script you should be able to see the logging in the ipython console.

Thank you very much. But I tried this, it does't work. The reson of this probelem maybe is the cooperation of tensorflow and Jupyter notebook. Because I can see the log information in the cmd.exe window which runs  jupyter.

**Comment by lyhue1991** [comment]
@facaiy Thank you very much. This limitation is really painful.

**Comment by facaiy** [comment]
@tomerk Hi, Tomer. Could you answer the question?

**Comment by tomerk** [comment]
Hi @facaiy what question still needs answering?
As you mentioned this is still a known limitation, because Jupyter notebooks don't capture C++ output.

We've decided to not add python-specific logic at this time (which we could do by running py_funcs internally) because it would prevent universal exporting of models that use `tf.print`

I've bumped the relevant issue on the ipython kernel's github:
https://github.com/ipython/ipykernel/issues/110

For anyone who needs a workaround there's a number of scattered code on the internet that tries to redirect C++ outputs to ipython, e.g.
https://github.com/minrk/wurlitzer

**Comment by facaiy** [comment]
Thanks for your detailed explanation, and workaround :-)

**Comment by lyhue1991** [comment]
@facaiy Yeah, maybe the best approach to solve this problem is to let ipykernel capture output coming from C and C++ libraries. @tomerk Thank you for your explanation.

**Comment by tomerk** [comment]
Just an update to this: We've actually come up with a solution to this inside of tensorflow, that should now be in the nightlies.

**Comment by Matoran** [comment]
When you use google colab, you can see the output in the runtime logs (runtime menu).
When you use jupyter lab (if it's your server) you need to check your terminal (from which you have launched jupyter lab).

**Comment by tomerk** [comment]
@Matoran is `tf.print` not currently printing in the cell output in both colab notebooks and jupyter notebooks? It should be, and if it's not we've had a regression and should fix it.

(Note that `tf.Print` in tf1 is not the same as `tf.print`)

**Comment by Matoran** [comment]
@tomerk I'm talking about `tf.Print` in tf1. `tf.print` works just fine. I had a bunch of pages talking about `tf.Print` and I answered here thinking it was another page. I will edit my text to avoid any confusion.
